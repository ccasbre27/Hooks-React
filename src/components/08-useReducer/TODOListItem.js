import React from 'react'
import PropTypes from 'prop-types';

export const TODOListItem = ({ index, todo, handleDelete, handleToggle }) => {
    return (
        <li 
            key={ todo.id }
            className="list-group-item">
            <p 
                className={ `${ todo.done && 'complete' }` }
                onClick={ () => handleToggle(todo.id) }> { index + 1 }. { todo.description } </p>
            <button
                className="btn btn-danger"
                onClick={ () => handleDelete(todo.id) }>
                    Delete
            </button>
        </li>
    )
}

TODOListItem.propTypes = {
    index: PropTypes.number.isRequired,
    todo: PropTypes.object.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired,
}