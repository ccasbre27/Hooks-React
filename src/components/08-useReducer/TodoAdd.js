import React from 'react'
import { useForm } from '../hooks/useForm';

export const TodoAdd = ({ handleAddTODO }) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if ( description.trim().length <= 1 ) {
            return;
        }

        const newTODO = {
            id: new Date().getTime(),
            description: description,
            done: false
        };

        handleAddTODO( newTODO );
        reset();

    }

    return (
        <div>
            <h4>Add TODO</h4>
            <hr />

            <form onSubmit={ handleSubmit }>
                <input 
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Learn..."
                    autoComplete="off" 
                    value={ description }
                    onChange={ handleInputChange } />

                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block">
                    Add
                </button>
            </form>
        </div>
    )
}
