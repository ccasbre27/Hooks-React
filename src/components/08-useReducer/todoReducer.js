
export const todoReducer = ( state = [], action ) => {

    switch ( action.type ) {
        
        case 'Add':
            return [ ...state, action.payload ];

        case 'Delete':
            // filter returns a new array without the one that match the criteria
            return state.filter( todo => todo.id !== action.payload )
    
        case 'Toggle':
            return state.map( todo => 
                ( todo.id === action.payload ) 
                    ? { ...todo, done: !todo.done }
                    : todo    
            )

        case 'Toggle-old':
            return state.map( todo => {
                if( todo.id === action.payload ) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                }
                else {
                    return todo;
                }
            });

        // it's been called the first time 
        default:
            return state;
    }

}