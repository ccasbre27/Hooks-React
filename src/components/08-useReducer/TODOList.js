import React from 'react'
import { TODOListItem } from './TODOListItem'
import PropTypes from 'prop-types';

export const TODOList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul className="list-group list-group-flush">
        {
            todos.map( (todo, i) => {
                return <TODOListItem 
                        index={ i }
                        todo={ todo }
                        handleDelete={ handleDelete }
                        handleToggle={ handleToggle } />
            })
        }
        </ul>
    )
}

TODOList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired,
}
