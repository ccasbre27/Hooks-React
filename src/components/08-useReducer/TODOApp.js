import React, { useEffect, useReducer } from 'react'
import { useForm } from '../hooks/useForm';
import './styles.css'
import { TodoAdd } from './TodoAdd';
import { TODOList } from './TODOList';
import { todoReducer } from './todoReducer';

const init = () => {
    // if there are not TODOs we return an empty array
    return JSON.parse(localStorage.getItem('TODOS')) || [];
}

export const TODOApp = () => {
    
    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('TODOS', JSON.stringify( todos ));
    }, [todos]);

    const handleDelete = ( todoId ) => {
        dispatch({
            type: 'Delete',
            payload: todoId
        });
    }

    const handleToggle = ( todoId ) => {
        dispatch({
            type: 'Toggle',
            payload: todoId
        });
    }

    const handleAddTODO = ( newTODO ) => {
        dispatch({
            type: 'Add',
            payload: newTODO
        });
    }
    
    return (
        <div>
            <h1>TODO App ({ todos.length })</h1>
            <hr />

            <div className="row">
                <div className="col-7">
                    <TODOList 
                        todos={ todos } 
                        handleDelete={ handleDelete } 
                        handleToggle={ handleToggle } />
                </div>
                <div className="col-5">
                    
                    <TodoAdd 
                        handleAddTODO={ handleAddTODO } />
                    
                </div>
            </div>
        </div>
    )
}
