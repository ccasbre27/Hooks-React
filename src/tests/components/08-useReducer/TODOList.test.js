
import { shallow } from 'enzyme';
import { TODOList } from './../../../components/08-useReducer/TODOList';
import { demoTodos } from './../../fixtures/demoTodos';

describe('Tests on <TODOList />', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    
    const wrapper = shallow( 
        <TODOList 
            todos={ demoTodos }
            handleDelete={ handleDelete }
            handleToggle={ handleToggle }
        />
    );

    test('should render the component', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('should have two <TODOListItem />', () => {
       
        expect( wrapper.find('TODOListItem').length ).toBe( demoTodos.length );
        expect( wrapper.find('TODOListItem').at( 0 ).prop( 'handleDelete' ) ).toEqual( expect.any(Function) );

    });
        
});
