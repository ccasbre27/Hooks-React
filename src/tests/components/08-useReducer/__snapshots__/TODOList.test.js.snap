// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <TODOList /> should render the component 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TODOListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    todo={
      Object {
        "description": "Lorem ipsum",
        "done": false,
        "id": 1,
      }
    }
  />
  <TODOListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    todo={
      Object {
        "description": "Bacon ipsum",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
